Scripts: Update webpack to v5 (try 2)

gziolo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Follow-up for #26382. 
Fixes #33768, fixes #32203, fixes #30289.

It's a second attempt to upgrade webpack to v5.

## Known Issues

- [x] `wp-scripts build` generates styles with a wrong name
  <img width="363" alt="Screen Shot 2021-08-02 at 15 15 49" src="https://user-images.githubusercontent.com/699132/127868546-a35d31bf-699b-4f6a-b782-8dcd2ea2b169.png">
 it should be `index-style.css` in the context of a block
- [x] `npm run build` errors because of API changes in webpack plugins
  <img width="1455" alt="Screen Shot 2021-08-02 at 15 23 15" src="https://user-images.githubusercontent.com/699132/127868792-1002a33f-14d4-4dc0-a924-e7cac241c773.png">
- [x] `npm run storybook:build` – Storybook has to be updated to work with webpack v5, more details at https://storybook.js.org/blog/storybook-for-webpack-5/

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

### Build for Gutenberg

- `npm run dev`
- `npm run build`

### Build for Storybook

- `npm run storybook:dev`
- `npm run storybook:build`

### Build for `@wordpress/scripts`

```bash
npx wp-create-block example --no-wp-scripts
cd example
../node_modules/.bin/wp-script build
```

## Types of changes
Breaking change for `@wordpress/scripts` users that provide their own config files for build tools.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

gziolo committed:
Scripts: Update webpack to v5 (try 2)

jsnajdr committed:
Remove the deprecated flatten option of CopyPlugin

jsnajdr commented:
I fixed the `copy-webpack-plugin` issue by removing the `flatten: true` option from its config. As we are always copying from a single directory, `build-style/*.css`, not from a tree, it should not be needed.

Another issue will be solved by merging #33785 and rebasing.

jsnajdr commented:
> `wp-scripts` build generates styles with a wrong name

@gziolo What's the fastest way to reproduce this?

gziolo commented:
> `wp-scripts build` generates styles with a wrong name

@gziolo What's the fastest way to reproduce this?

The steps I use to create a custom block:

```bash
npx wp-create-block example --no-wp-scripts
cd example
../node_modules/.bin/wp-script build
```

jsnajdr commented:
> The steps I use to create a custom block:

Thanks, now I have the `647.css` file, too 🙂

walbo commented:
I looked into updating to v5 when working on #33785. I belive `@wordpress/custom-templated-path-webpack-plugin` and `@wordpress/library-export-default-webpack-plugin` that has a few issues in v5 can be removed. You now have the option to set `library` in the entry config that will fix what those packages do. If you want I can create a PR that targets this branch.

Commit:
Try: Webpack 5 packages entry library config

gziolo commented:
> I looked into updating to v5 when working on #33785. I belive @wordpress/custom-templated-path-webpack-plugin and @wordpress/library-export-default-webpack-plugin that has a few issues in v5 can be removed. You now have the option to set library in the entry config that will fix what those packages do.

If it is possible then it would be great and we could stop maintaining those two custom plugins. They were introduced only to deal with webpack's limitations.

walbo commented:
> If it is possible then it would be great and we could stop maintaining those two custom plugins. They were introduced only to deal with webpack's limitations.

Created #33826 based on this branch that removed those plugins. Runs `npm run build` without errors (as long as you rebase on trunk so #33785 is included)

jsnajdr committed:
Custom name function for split chunks

jsnajdr commented:
The `647.css` issue is solved. I had to provide a custom `name` function for the `splitChunks.cacheGroup` config that produced the desired name for the asset: cache group name (`style`) + original chunk name (`index`). Webpack 5 algorithm for splitting chunks and naming them is more complicated than it was in version 4.

gziolo committed:
Update the default webpack config for `@wordpress/scripts`
It adjusts handling for assets that changed in v5.

jsnajdr commented:
@walbo The changes in #33826 look good and are very helpful, so I cherry-picked the commit to this branch. That makes things simpler to administer.

jsnajdr commented on tools/webpack/packages.js:
```diff
@@ -31,44 +29,47 @@ const gutenbergPackages = Object.keys( dependencies )
+	'server-side-render',
+	'shortcode',
+	'warning',
+];
```
It would be really nice if this information was part of the individual packages, like a flag in `packages/api-fetch/package.json`. The list shoudn't be maintained in a build script.

It's certainly a task for another PR, it just caught my attention here and I wanted to make a note.

gziolo replied:
I agree that it's a bit unfortunate that we have to list all of them. The benefit would be that we wouldn't have to maintain the same list in WordPress core. Still, it's a manual task that we will have to remember about.

jsnajdr committed:
Update hashes in test snapshots

gziolo committed:
Deprecate Library Export Default Webpack Plugin for webpack 5

gziolo commented:
All unit tests should be fixed with https://github.com/WordPress/gutenberg/pull/33818/commits/abbc15ff58a21e0e4d41fcdb8f4841ea541eeef5. I marked `@wordpress/library-export-default-webpack-plugin` as deprecated in the context of webpack 5 and removed unit tests that were erroring. Unless you feel we should try to fix them and keep the plugin compatible with webpack 5. As discussed, it's obsolete now so better to ignore it.

In https://github.com/WordPress/gutenberg/pull/33818/commits/b257e9e3f871274ec0bcb564d876f9c83b74a22e, I added some tweaks to the default webpack config shipped with `@wordpress/scripts`. The biggest change is that `file-loader` is no longer necessary because webpack 5 has native assets handling. I tested fonts handling, I still need to verify images, including SVG.

gziolo commented on packages/readable-js-assets-webpack-plugin/test/fixtures/webpack.config.js:
```diff
@@ -17,9 +17,7 @@ module.exports = {
 		concatenateModules: true,
 		minimizer: [
 			new TerserPlugin( {
-				cache: true,
```
@jsnajdr, is it correct? I had to remove both flags because they were removed in webpack 5. The output generated in the tests needs to be verified, too.

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -135,16 +104,16 @@ const config = {
 		},
 		minimizer: [
 			new TerserPlugin( {
-				cache: true,
 				parallel: true,
-				sourceMap: ! isProduction,
```
A similar concern as earlier, both flags were removed from the plugin to use settings from webpack. It would be great to verify if we need to add more changes.

gziolo commented on packages/readable-js-assets-webpack-plugin/test/fixtures/webpack.config.js:
```diff
@@ -17,9 +17,7 @@ module.exports = {
 		concatenateModules: true,
 		minimizer: [
 			new TerserPlugin( {
-				cache: true,
 				parallel: true,
-				sourceMap: true,
```
@jsnajdr, is it correct? I had to remove both flags because they were removed in webpack 5. The output generated in the tests needs to be verified, too.

jsnajdr replied:
Yes, removing both flags is correct IMO. According to the `terser-webpack-plugin` [5.0.0 release notes](https://github.com/webpack-contrib/terser-webpack-plugin/releases/tag/v5.0.0), these options are now inherited from the parent webpack config.

gziolo commented on tools/webpack/shared.js:
```diff
@@ -19,15 +19,14 @@ const {
 			mode === 'production' && ! process.env.WP_BUNDLE_ANALYZER,
 		minimizer: [
 			new TerserPlugin( {
-				cache: true,
 				parallel: true,
-				sourceMap: mode !== 'production',
```
Another place where `TerserPlugin` had changes in config enforced by the API changes.

jsnajdr commented:
> I marked `@wordpress/library-export-default-webpack-plugin` as deprecated in the context of webpack 5 and removed unit tests that were erroring.

Are we going to remove the legacy plugins from the repo? We won't do any further releases of them, will we? If we did, it would be preferrable to keep the unit tests working.

jsnajdr commented:
@gziolo There are also unit tests failing in the `readable-js-assets-webpack-plugin`. I'm working on fixing them.

gziolo committed:
Update `@wordpress/scripts` changelog file

gziolo committed:
Docs: Add changelog entry in `@wordpress/dependency-extraction-webpack-plugin`

gziolo committed:
Revert `webpack-sources` version upgrade

gziolo commented:
> > I marked @wordpress/library-export-default-webpack-plugin as deprecated in the context of webpack 5 and removed unit tests that were erroring.
> 
> Are we going to remove the legacy plugins from the repo? We won't do any further releases of them, will we? If we did, it would be preferrable to keep the unit tests working.

I was thinking about publishing a new version with updates in the readme and the changelog. The next step could be deprecating on npm the package and removing the source folder. However, if you know how to fix the tests then we can keep the sources.

jsnajdr commented:
> However, if you know how to fix the tests then we can keep the sources.

I'll see what I can do.

gziolo committed:
Make Storybook run with webpack 5

gziolo commented:
`npm run storybook:build` works with webpack 5. I still need to update some less important config entries that I temporarily commented out and verify everything with `npm run storybook:dev`.

jsnajdr committed:
Log asset names in test messages, avoid optimizing out the entire source

gziolo commented on storybook/webpack.config.js:
```diff
@@ -10,11 +10,11 @@ const postcssPlugins = require( '@wordpress/postcss-plugins-preset' );
 module.exports = ( { config } ) => {
 	config.module.rules.push(
-		{
+		/*{
```
Details on how it should be configured in https://www.npmjs.com/package/@storybook/source-loader.

gziolo committed:
Improve webpack config for Storybook

gziolo commented:
Story source code view works again:

<img width="1055" alt="Screen Shot 2021-08-03 at 16 08 30" src="https://user-images.githubusercontent.com/699132/128032214-63ac6962-4aa1-4baa-ab1c-d6bb268224cc.png">

I tested both `npm run storybook:dev` and `npm run storybook:build`.

walbo commented:
Nice work!

`engines` in package.json should be updated to `>=12.13`. Some of the updated packages require node v12.13.0. (ex [mini-css-extract-plugin v2.0.0](https://github.com/webpack-contrib/mini-css-extract-plugin/releases/tag/v2.0.0), [css-loader v6.0.0)](https://github.com/webpack-contrib/css-loader/releases/tag/v6.0.0)

gziolo commented:
> `engines` in package.json should be updated to `>=12.13`. Some of the updated packages require node v12.13.0. (ex [mini-css-extract-plugin v2.0.0](https://github.com/webpack-contrib/mini-css-extract-plugin/releases/tag/v2.0.0), [css-loader v6.0.0)](https://github.com/webpack-contrib/css-loader/releases/tag/v6.0.0)

Good catch, let's align 👍🏻 Done in https://github.com/WordPress/gutenberg/pull/33818/commits/032456862578b797a1f431237f3b0cf2c0e54a88.

gziolo committed:
Increase the minimum Node version in `@wordpress/scripts`

jsnajdr commented:
> However, if you know how to fix the tests then we can keep the sources.

@gziolo The `library-export-default-webpack-plugin` is not going to work with webpack 5 without major changes. Currently, it works by appending a string suffix `["default"]` to the module source string, assuming that the module source has certain layout: that it is an expression that's indexable with the `[]` operator. But in webpack 5, the source layout is different.

Simply removing the tests is the best way to proceed, I'm now confirming that.

jsnajdr reviewed (approved):
Everything I tested works well and I don't see any further obstacles. Looks good to go :+1:

gziolo commented:
I'm working on a test block template that we can use with `@wordpress/create-block` to compare the build output generated with webpack 4 and webpack 5. We will have to confirm that:

- [x] file names generated are the same
- [x] CSS integration works correctly
- [x] SVG integration still works
- [x] image integration still works
- [x] font integration still works

I need to finish updating [wp-block-directory-template](https://www.npmjs.com/package/wp-block-directory-template) package so we can scaffold a block and compare it as follows:

## webpack 4 (`@wordpress/scripts` v17.1.0)

```bash
npx @wordpress/create-block example --template wp-block-directory-template
cd example
# npm run build is executed as part of the scaffolding process
npm start
```

## webpack 5 (this branch)

```bash
npx wp-create-block example --no-wp-scripts --template wp-block-directory-template
cd example
../node_modules/.bin/wp-script build
../node_modules/.bin/wp-script start
```

jsnajdr commented:
OK, please ping me whenever I can contribute some debugging 🙂

gziolo commented:
I tested again both `npm run build` and `npm run dev` and everything works perfectly fine. I can also confirm that updates to CSS and JS source files in the watch mode trigger updates to all build folders as before.

gziolo commented:
I discovered an issue in webpack v5 with how it handles `wp-script start` in comparison with webpack v4.

webpack 4:
<img width="296" alt="Screen Shot 2021-08-04 at 14 29 27" src="https://user-images.githubusercontent.com/699132/128181137-1865d2d5-8302-4b84-a4bd-f62bae9d3e10.png">

webpack 5 - additional `style-index.js` file that breaks the block:
<img width="278" alt="Screen Shot 2021-08-04 at 14 29 38" src="https://user-images.githubusercontent.com/699132/128181142-f12f9ae3-3c67-4b34-b560-f9de2aa7ff18.png">
 
`wp-scripts build` works the same way with webpack v4 and v5.

We had a custom plugin that was removing `style-index.js` in webpack v4, so we need to investigate how it can be handled natively in v5 since it isn't an issue for `wp-scripts build`.

gziolo committed:
Ensure sthere is no empty JS file for exctracted CSS file

gziolo commented:
b102916 fixes the issue with obsolete `style-index.js` generated with `wp-scripts start`. The solution I used is now included in the official docs and I found it in the following comment: https://github.com/webpack/webpack/issues/7300#issuecomment-702840962. The related issue was https://github.com/webpack-contrib/mini-css-extract-plugin/issues/85 - we had it referenced in the config file.

gziolo commented:
I can also confirm that SVG and images integration works as expected:
- SVG files are inlined a

Source code used for testing:

<img width="1363" alt="Screen Shot 2021-08-04 at 15 50 13" src="https://user-images.githubusercontent.com/699132/128193931-0e40950f-0230-47aa-87a6-099845bc8499.png">
<img width="869" alt="Screen Shot 2021-08-04 at 15 57 05" src="https://user-images.githubusercontent.com/699132/128193940-ac5ae9c8-0541-4b93-bbf6-d513b5c5da81.png">

Project structure:

<img width="268" alt="Screen Shot 2021-08-04 at 15 57 44" src="https://user-images.githubusercontent.com/699132/128193984-8d9879ea-26d3-4f66-8d3c-f160203c5fcc.png">

SVG file was inlined in HTML output:

<img width="1052" alt="Screen Shot 2021-08-04 at 16 00 40" src="https://user-images.githubusercontent.com/699132/128194485-c0be5b39-c437-4f04-8d59-a00d0f7b2391.png">


Inserted block:

<img width="664" alt="Screen Shot 2021-08-04 at 15 57 13" src="https://user-images.githubusercontent.com/699132/128194052-61cf8cb9-efc0-4bd0-885e-e19352500f9e.png">

Mamaduka commented:
It would be interesting to see how this affects plugins that use `@wordpress/scripts` but also extend default webpack config.

gziolo commented:
> It would be interesting to see how this affects plugins that use @wordpress/scripts but also extend default webpack config.

I'm sure it won't be simple to upgrade for them. We made it clear in the documentation that using own webpack config is discouraged. See https://github.com/WordPress/gutenberg/tree/trunk/packages/scripts#extending-the-webpack-config:

> If you follow this approach, please, be aware that future versions of this package may change what webpack and Babel plugins we bundle, default configs, etc. Should those changes be necessary, they will be registered in the package’s CHANGELOG, so make sure to read it before upgrading.

gziolo commented:
I published a new version of `@wordpress/scripts` to npm to make it possible to test the upcoming changes. It is hidden behind the `next` dist tag:

```bash
npm i --save-dev @wordpress/scripts@next
```

gziolo commented:
I see an error that needs to be resolved before the final release:

```
[webpack-cli] Error: No browserslist config found to handle the 'browserslist' target.
See https://github.com/browserslist/browserslist#queries for possible ways to provide a config.
The recommended way is to add a 'browserslist' key to your package.json and list supported browsers (resp. node.js versions).
You can also more options via the 'target' option: 'browserslist' / 'browserslist:env' / 'browserslist:query' / 'browserslist:path-to-config' / 'browserslist:path-to-config:env'
```

It looks like we need to provide the fallback path to the browserlist config with `browserslist:/path/to/config` when a project doesn't provide its own config.

gziolo commented:
Fix is ready: #34112.

gziolo commented:
Follow-up issues reported:
- #34236
- #34239
